File: algorithm_leetcode/src/main/python/com/huldar/python/day06/ThreeNumSum.py
#!usr/bin/env python
# -*- coding:utf-8 _*-
"""
@author:huldar
@file: ThreeNumSum.py
@time: 2019/07/22
"""
from typing import List


class Solution:
    @staticmethod
    def three_num_sum(nums: List[int]) -> List[List[int]]:
        """
        三数之和.求三个数的和为0的组合
        :param nums:
        :return:
        """
        if len(nums) < 3: return []
        nums.sort()
        result = []
        i = 0
        while i < len(nums) - 2:
            dic = []
            if nums[i] > 0: break
            if i > 0 and nums[i] == nums[i - 1]:
                i += 1
                continue
            for j in range(i + 1, len(nums)):
                if nums[j] in dic and dic[nums[j]][2] == 0:
                    dic[nums[j]][2] = 1
                    if j > i + 2 and nums[j] == nums[j - 2]:
                        continue
                    result.append(dic[nums[j]][:2] + [nums[j]])
                else:
                    dic[-(nums[i] + nums[j])] = [nums[i], nums[j], 0]
            i += 1
        return result

File: algorithm_leetcode/src/main/java/com/huldar/java/day06/ThreeNumSum.java
package com.huldar.java.day06;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * 三数之和
 *
 * @author huldar
 * @date 2019/6/18 16:28
 */
public class ThreeNumSum {
    public static List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> result = new ArrayList<>();
        //如果原始数组为null或者其长度小于3 直接返回
        if (null == nums || nums.length < 3) {
            return result;
        }
        //从小到大进行排序
        Arrays.sort(nums);
        int length = nums.length;
        for (int i = 0; i < length; i++) {
            //如果当前数大于0,则三数之和一定大于0
            if (nums[i] > 0) {
                break;
            }
            //去重
            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }
            //双指针
            int left = i + 1;
            int right = length - 1;
            while (left < right) {
                int sum = nums[i] + nums[left] + nums[right];
                if (sum == 0) {
                    result.add(Arrays.asList(nums[i], nums[left], nums[right]));
                    //去重
                    while (left < right && nums[left] == nums[left + 1]) {
                        left++;
                    }
                    //去重
                    while (left < right && nums[right] == nums[right - 1]) {
                        right--;
                    }
                    left++;
                    right--;
                } else if (sum < 0) {
                    left++;
                } else {
                    right--;
                }
            }
        }
        return result;
    }
}
